core/services/ocr2/plugins/generic/telemetry_adapter.go

```diff
@@ -0,0 +1,54 @@
+package generic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+)
+
+var _ types.TelemetryService = (*TelemetryAdapter)(nil)
+
+type TelemetryAdapter struct {
+	endpointGenerator types.MonitoringEndpointGenerator
+	endpoints         map[string]commontypes.MonitoringEndpoint
+}
+
+func (t *TelemetryAdapter) Send(ctx context.Context, network string, chainID string, contractID string, telemetryType string, payload []byte) error {
+	e, err := t.getOrCreateEndpoint(contractID, telemetryType, network, chainID)
+	if err != nil {
+		return err
+	}
+	e.SendLog(payload)
+	return nil
+}
+
+func (t *TelemetryAdapter) getOrCreateEndpoint(contractID string, telemetryType string, network string, chainID string) (commontypes.MonitoringEndpoint, error) {
+	if contractID == "" {
+		return nil, errors.New("contractID cannot be empty")
+	}
+	if telemetryType == "" {
+		return nil, errors.New("telemetryType cannot be empty")
+	}
+	if network == "" {
+		return nil, errors.New("network cannot be empty")
+	}
+	if chainID == "" {
+		return nil, errors.New("chainID cannot be empty")
+	}
+
+	key := makeKey(network, chainID, contractID, telemetryType)
```

Comment:
Could use an array for these keys :shrug: 
```suggestion
	key := [4]string{network, chainID, contractID, telemetryType}
```

Reply:
Doh, yes this is much better!